dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -151,20 +151,24 @@ private String getToken(List<Token> tokens, TokenType tokenType) {
      * workflow verions.
      * @param workflow    workflow to be updated
      * @param newWorkflow workflow to grab new content from
+     * @param user
+     * @param versionName
      */
-    protected void updateDBWorkflowWithSourceControlWorkflow(Workflow workflow, Workflow newWorkflow, final User user) {
+    protected void updateDBWorkflowWithSourceControlWorkflow(Workflow workflow, Workflow newWorkflow, final User user, Optional<String> versionName) {
         // update root workflow
         workflow.update(newWorkflow);
         // update workflow versions
         Map<String, WorkflowVersion> existingVersionMap = new HashMap<>();
         workflow.getWorkflowVersions().forEach(version -> existingVersionMap.put(version.getName(), version));
 
-        // delete versions that exist in old workflow but do not exist in newWorkflow
-        Map<String, WorkflowVersion> newVersionMap = new HashMap<>();
-        newWorkflow.getWorkflowVersions().forEach(version -> newVersionMap.put(version.getName(), version));
-        Sets.SetView<String> removedVersions = Sets.difference(existingVersionMap.keySet(), newVersionMap.keySet());
-        for (String version : removedVersions) {
-            workflow.removeWorkflowVersion(existingVersionMap.get(version));
+        // delete versions that exist in old workflow but do not exist in newWorkflow (only for whole refresh)
+        if (versionName.isEmpty()) {
```

Comment:
What happens if versionName is not empty and the version got deleted on GitHub?

Reply:
Good call. Looks like the endpoint will do nothing and not throw an error. I'll update it so that it throws an error if the branch does not exist.

Reply:
Actually think I misunderstood. Do you mean if the refresh is of a version that doesn't exist on GitHub but exists on Dockstore, it should be deleted on Dockstore?

Reply:
After discussing with Denis, we've decided that on refresh of a version, if that version does not exist, throw an error. This is different from the current refresh workflow logic, which would delete a version on Dockstore if it doesn't exist on the Git repo anymore. We decided on this difference because of how we expect the users to use this refresh in the UI. If they are refreshing a version, then likely they expect it to still exist and are not trying to delete it. A delete version button may be present in the future, pending archive/delete/life cycle discussions.

Reply:
> Actually think I misunderstood. Do you mean if the refresh is of a version that doesn't exist on GitHub but exists on Dockstore, it should be deleted on Dockstore?

Yes, I meant that. :) Although I think I missed the larger context, and was just assuming it was a deleted version in general, as opposed to the specific version you are refreshing. You have it covered with the other ticket in any case.